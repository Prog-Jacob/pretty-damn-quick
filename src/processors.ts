#!/usr/bin/env node

import * as fs from "fs";
import * as path from "path";
import prettier from "prettier";
import micromatch from "micromatch";

import {
  getDiffFileList,
  getDiffForFile,
  getRangesForDiff,
  getUntrackedFileList,
  hasCleanIndex,
  fetchFromOrigin,
} from "./git";
import { guessBranch } from "./ci";
import log from "./log";
import { LineOffsets } from "./offsets";
import { insertMarkers, mergeMarkedSections } from "./marker";

// ================================
// 1. Branch Fetching for CI
// ================================
// When running in CI, ensure we fetch the latest branch for diff computation.
if (process.env.CI !== undefined) {
  const branch = process.env.ESLINT_PLUGIN_DIFF_COMMIT ?? guessBranch();

  if (branch !== undefined) {
    const branchWithoutOrigin = branch.replace(/^origin\//, "");
    const branchWithOrigin = `origin/${branchWithoutOrigin}`;

    fetchFromOrigin(branchWithoutOrigin);
    process.env.ESLINT_PLUGIN_DIFF_COMMIT = branchWithOrigin;
  }
}

// ================================
// 2. CLI Options Interface
// ================================
type PrettierOptionsCLI = {
  check: boolean; // --check: only check formatting
  staged: boolean; // --staged: only staged files
  changed: boolean; // --changed: only changed files
  trackedOnly: boolean; // --tracked-only: doesn't process untracked files
  lines: boolean | "experimental"; // --lines: format only changed lines, or 'experimental'
  extensions?: string[]; // e.g., ['ts','js']: filter by file extensions
  pattern?: string; // glob pattern to filter files
};

// ================================
// 3. Entry Point
// ================================
async function runPrettier(options: PrettierOptionsCLI): Promise<void> {
  const targetFiles = resolveTargetFiles(options);

  if (targetFiles.length === 0) {
    log.info("No files to process.");
    return;
  }

  let allOk = true;
  let hasExitingError = false;

  for (const file of targetFiles) {
    try {
      const ok =
        options.lines === "experimental"
          ? await processRangesWithMarkers(file, options)
          : options.lines
            ? await processFileByRanges(file, options)
            : await processWholeFile(file, options);
      if (!ok) {
        allOk = false;
      }
    } catch (err) {
      log.error(err, file);

      const errMessage =
        err instanceof Error ? (err.stack ?? err.message) : String(err);

      if (err instanceof SyntaxError || !errMessage.includes("prettier")) {
        hasExitingError = true;
      }
    }
  }

  if (hasExitingError) {
    throw new Error(
      "Prettier exiting error(s) occurred. See above for details.",
    );
  } else if (options.check && !allOk) {
    throw new Error("Some files are not formatted.");
  }

  process.exitCode = 0;
}

// ================================
// 4. Resolve Target Files
// ================================
function resolveTargetFiles(options: PrettierOptionsCLI): string[] {
  const stagedFiles: string[] = getDiffFileList(true);
  const changedFiles: string[] = getDiffFileList(false);
  const untrackedFiles: string[] = getUntrackedFileList(false);

  let files: string[] = [];

  if (options.staged === options.changed || options.staged) {
    files.push(...stagedFiles);
  }

  if (options.staged === options.changed || options.changed) {
    files.push(...changedFiles);
  }

  if (!options.trackedOnly) {
    files.push(...untrackedFiles);
  }

  // Remove duplicates
  files = Array.from(new Set(files));

  // Filter by extensions if provided
  if (options.extensions?.length) {
    const exts = options.extensions ?? [];
    files = files.filter((file) => exts.includes(path.extname(file).slice(1)));
  }

  files = files.filter(
    (file) =>
      fs.statSync(file).isFile() && !path.basename(file).startsWith("."),
  );

  // Filter by glob pattern if provided (from positional argument)
  if (options.pattern) {
    files = micromatch.match(files, options.pattern, { dot: true });
  }

  return files;
}

// ================================
// 5. Process Whole File
// ================================
async function processWholeFile(
  file: string,
  options: PrettierOptionsCLI,
): Promise<boolean> {
  // Skip files with unstaged changes if in staged mode
  if (options.staged && !hasCleanIndex(file)) {
    log.skipped(
      file,
      "has unstaged changes. Please stage or remove the changes.",
    );
    return true;
  }

  const code = fs.readFileSync(file, "utf-8");
  const info = await prettier.getFileInfo(file);
  const config = await prettier.resolveConfig(file);

  if (info.ignored || info.inferredParser === null) {
    return true;
  }

  const formatted = await prettier.format(code, {
    ...(config ?? {}),
    filepath: file,
  });
  const isNoChange = code.length === formatted.length && code === formatted;

  if (options.check) {
    if (!isNoChange) {
      log.checked(file);
      return false;
    }
  } else if (!isNoChange) {
    fs.writeFileSync(file, formatted, "utf-8");
    log.formatted(file);
  }

  return isNoChange;
}

// ================================
// 6. Process File by Changed Ranges
// ================================

async function processFileByRanges(
  file: string,
  options: PrettierOptionsCLI,
): Promise<boolean> {
  const info = await prettier.getFileInfo(file);
  const config = await prettier.resolveConfig(file);

  if (info.ignored || info.inferredParser === null) {
    return true;
  }

  const originalText = fs.readFileSync(file, "utf-8");
  const lineOffsets = new LineOffsets(originalText);
  const diff = getDiffForFile(file, options.staged);
  const ranges = getRangesForDiff(diff);

  if (
    ranges.length === 1 &&
    ranges[0]?.rangeStart() === 0 &&
    Math.abs(ranges[0]?.rangeEnd() - lineOffsets.totalLines()) <= 1
  ) {
    return processWholeFile(file, options);
  }

  let currentText = originalText;

  for (const range of ranges) {
    const startLine = range.rangeStart();
    const endLine = range.rangeEnd();
    const start = lineOffsets.getOffset(startLine);
    const end = lineOffsets.getOffset(endLine) - 1;

    currentText = await prettier.format(currentText, {
      ...(config ?? {}),
      filepath: file,
      rangeStart: start,
      rangeEnd: end,
    });

    if (options.check) {
      const originalSegment = originalText.slice(start, end);
      const currentSegment = currentText.slice(start, end);
      const isNoChange =
        originalSegment.length === currentSegment.length &&
        originalSegment === currentSegment;

      if (!isNoChange) {
        log.checked(file, `${startLine + 1}-${endLine}`);
      }
    }
  }

  const isNoChange =
    currentText.length === originalText.length && currentText === originalText;

  if (!options.check && !isNoChange) {
    fs.writeFileSync(file, currentText, "utf-8");
  }

  return isNoChange;
}

// ================================
// 7. Process File by Changed Ranges Using Markers
// ================================

async function processRangesWithMarkers(
  file: string,
  options: PrettierOptionsCLI,
): Promise<boolean> {
  const originalText = fs.readFileSync(file, "utf-8");
  const info = await prettier.getFileInfo(file);
  const config = await prettier.resolveConfig(file);

  if (info.ignored || info.inferredParser === null) {
    return true;
  }

  const diff = getDiffForFile(file, options.staged);
  const ranges = getRangesForDiff(diff);

  if (ranges.length === 1 && ranges[0]?.rangeStart() === 0) {
    const lineOffsets = new LineOffsets(originalText);

    if (Math.abs(ranges[0]?.rangeEnd() - lineOffsets.totalLines()) <= 1) {
      return processWholeFile(file, options);
    }
  }

  try {
    const markedText = insertMarkers(originalText, ranges, info.inferredParser);
    const formattedWithMarkers = await prettier.format(markedText, {
      ...(config ?? {}),
      filepath: file,
    });
    const mergedText = mergeMarkedSections(
      markedText,
      formattedWithMarkers,
      info.inferredParser,
    );

    if (options.check) {
      if (mergedText !== originalText) {
        for (const range of ranges) {
          log.checked(file, `${range.rangeStart() + 1}-${range.rangeEnd()}`);
        }
        return false;
      }
    } else if (mergedText !== originalText) {
      fs.writeFileSync(file, mergedText, "utf-8");
      log.formatted(file);
    }

    return true;
  } catch (error) {
    log.error(error, file);

    // Fallback to line-based processing if marker-based fails
    return processFileByRanges(file, options);
  }
}

// Top-level export
export {
  runPrettier,
  resolveTargetFiles,
  processWholeFile,
  processFileByRanges,
  processRangesWithMarkers,
};
export type { PrettierOptionsCLI };
