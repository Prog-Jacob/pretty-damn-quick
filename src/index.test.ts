process.env.CI = "true";
import * as child_process from "child_process";
import { type PrettierOptionsCLI } from "./processors";

jest.mock("child_process");
const mockedChildProcess = jest.mocked(child_process, { shallow: true });
mockedChildProcess.execFileSync.mockReturnValue(
  Buffer.from("line1\nline2\nline3"),
);

type YargsMock = {
  options: jest.Mock;
  help: jest.Mock;
  parseSync: jest.Mock;
};

const createYargsMock = (): YargsMock =>
  Object.assign(jest.fn(), {
    options: jest.fn().mockReturnThis(),
    help: jest.fn().mockReturnThis(),
    parseSync: jest.fn(),
  });

const yargsMock = createYargsMock();

jest.mock("yargs", () => ({
  __esModule: true,
  default: jest.fn(() => yargsMock),
}));

jest.mock("yargs/helpers", () => ({
  hideBin: jest.fn((argv: string[]) => argv),
}));

let runCli: (argv: string[]) => Promise<void>;

beforeAll(async () => {
  const mod = await import("./index");
  runCli = mod.runCli;
});

describe("CLI entrypoint", () => {
  let runPrettierSpy: jest.SpyInstance<Promise<void>, [PrettierOptionsCLI]>;

  beforeEach(async () => {
    const processors = await import("./processors");
    runPrettierSpy = jest
      .spyOn(processors, "runPrettier")
      .mockResolvedValue(undefined);
  });

  afterEach(() => {
    jest.clearAllMocks();
    process.exitCode = 0;
  });

  it("parses CLI args and calls runPrettier with correct options", async () => {
    yargsMock.parseSync.mockReturnValue({
      check: true,
      staged: true,
      extensions: "ts,js",
    });

    await runCli([
      "node",
      "index.js",
      "--check",
      "--staged",
      "--extensions",
      "ts,js",
    ]);

    expect(runPrettierSpy).toHaveBeenCalledWith({
      check: true,
      staged: true,
      changed: false,
      lines: false,
      extensions: ["ts", "js"],
    });
  });

  it("defaults options when none provided", async () => {
    yargsMock.parseSync.mockReturnValue({});

    await runCli(["node", "index.js"]);

    expect(runPrettierSpy).toHaveBeenCalledWith({
      check: false,
      staged: false,
      changed: false,
      lines: false,
      extensions: [],
    });
  });

  it("calls process.exit(1) and logs error if runPrettier throws", async () => {
    const error = new Error("fail!");
    const processors = await import("./processors");
    jest.spyOn(processors, "runPrettier").mockRejectedValueOnce(error);

    const errorSpy = jest
      .spyOn(console, "error")
      .mockImplementation(() => undefined);

    yargsMock.parseSync.mockReturnValue({ check: true });

    await runCli(["node", "index.js", "--check"]);

    expect(errorSpy).toHaveBeenCalledWith(
      "Error: " + (error.message ?? String(error)),
    );
    expect(process.exitCode).toBe(1);

    errorSpy.mockRestore();
  });
});

describe("exports", () => {
  it("exports expected runPrettier function", async () => {
    jest.resetModules();
    const real = await import("./processors");
    expect(typeof real.runPrettier).toBe("function");
    expect(real.runPrettier).toMatchSnapshot();
  });

  it("allows direct runPrettier call through index (spy only)", async () => {
    jest.resetModules();
    const processors = await import("./processors");
    const spy = jest
      .spyOn(processors, "runPrettier")
      .mockResolvedValue(undefined);
    const { runPrettier } = await import("./index");

    await expect(
      runPrettier({
        check: false,
        staged: false,
        changed: false,
        lines: false,
        extensions: [],
      }),
    ).resolves.toBeUndefined();

    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
});
