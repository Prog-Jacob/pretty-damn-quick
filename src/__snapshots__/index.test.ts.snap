// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exports exports expected runPrettier function 1`] = `[Function]`;
